exonum/src/events/noise/mod.rs

```diff
@@ -52,16 +52,17 @@ fn listen_handshake(stream: TcpStream, params: &HandshakeParams) -> HandshakeRes
     let max_message_len = params.max_message_len;
     let mut noise = NoiseWrapper::responder(params);
     let framed = read(stream).and_then(move |(stream, msg)| {
-        let _buf = noise.read_handshake_msg(&msg);
-        write_handshake_msg(&mut noise)
-            .and_then(|(len, buf)| write(stream, &buf, len))
-            .and_then(|(stream, _msg)| read(stream))
-            .and_then(move |(stream, msg)| {
-                let _buf = noise.read_handshake_msg(&msg);
-                let noise = noise.into_transport_mode()?;
-                let framed = stream.framed(MessagesCodec::new(max_message_len, noise));
-                Ok(framed)
-            })
+        read_handshake_msg(&msg, &mut noise).and_then(move |_| {
```

Comment:
Slightly off-topic: Would it be possible to refactor `read_handshake_msg` / `write_handshake_msg` so that they deal directly with streams? For now, they introduce boilerplate (`write` after `write_handshake_msg`, `read` before `read_handshake_msg` and are not really asynchronous.

Reply:
Good idea, but it's difficult to pass `noise` into the future because it needs to have static lifetime. It also doesn't implement the `Clone` trait. For example:

```rust
fn read_handshake_msg_stream(&self,
    stream: TcpStream,
    noise: &'static mut NoiseWrapper,
) -> Box<Future<Item = (usize, Vec<u8>), Error = io::Error>> {
    let res = read(stream).and_then(move |(stream, msg)| {
        noise.read_handshake_msg(&msg).map_err(|e| e.into())
    });

    Box::new(res)
}
```
Maybe it will be possible using `RefCell` of something, but I don't think it will be more readable.

Reply:
I've had [something like this](https://gist.github.com/slowli/cb3bd293744d5f97287490da791e2ff4) in mind (it compiles, but I skipped the tests). In particular, the code like

```rust
    let framed = channel.read_handshake_msg(stream)
        .and_then(|(stream, channel)| channel.write_handshake_msg(stream))
        .and_then(|(stream, channel)| channel.read_handshake_msg(stream))
        .and_then(|(stream, channel)| channel.finalize(stream));
```
looks much more discernible to me than the current implementation.

Now that I think about it, it may make sense to mock `Handshake`, not `HandshakeChannel` (i.e., make `read_handshake_msg` / `write_handshake_msg` internals of `NoiseHandshake`, and get rid of `HandshakeChannel` altogether).